src/endless/EndlessUsbToolDlg.cpp

```diff
@@ -5323,3 +5333,267 @@ BOOL CEndlessUsbToolDlg::IsBitlockedDrive(const CString &drive)
+			IFFALSE_PRINTERROR(result == ERROR_SUCCESS, "Error on setting key to original value.");
+		}
+		break;
+	default:
+		uprintf("ERROR: registry key type %d not handled", keyType);
+		goto error;
+	}
+
+done:
+	retResult = TRUE;
+error:
+	return retResult;
+
+}
+
+BOOL CEndlessUsbToolDlg::MountESPFromDrive(HANDLE hPhysical, const char **espMountLetter, const CString &systemDriveLetter)
+{
+	FUNCTION_ENTER;
+
+	BOOL retResult = FALSE;
+	BYTE layout[4096] = { 0 };
+	PDRIVE_LAYOUT_INFORMATION_EX DriveLayout = (PDRIVE_LAYOUT_INFORMATION_EX)(void*)layout;
+	DWORD size;
+	BOOL result;
+
+	// get partition layout
+	result = DeviceIoControl(hPhysical, IOCTL_DISK_GET_DRIVE_LAYOUT_EX, NULL, 0, layout, sizeof(layout), &size, NULL);
+	IFFALSE_GOTOERROR(result != 0 && size > 0, "Error on querying disk layout.");
+	IFFALSE_GOTOERROR(DriveLayout->PartitionStyle == PARTITION_STYLE_GPT, "Unexpected partition type. Partition style is not GPT");
+
+	DWORD efiPartitionNumber = -1;
+	PARTITION_INFORMATION_EX *partition = NULL;
+	for (DWORD index = 0; index < DriveLayout->PartitionCount; index++) {
+		partition = &(DriveLayout->PartitionEntry[index]);
+
+		if (partition->Gpt.PartitionType == PARTITION_SYSTEM_GUID) {
+			uprintf("Found ESP\r\nPartition %d:\r\n  Type: %s\r\n  Name: '%ls'\r\n ID: %s",
+				index + 1, GuidToString(&partition->Gpt.PartitionType), partition->Gpt.Name, GuidToString(&partition->Gpt.PartitionId));
+			efiPartitionNumber = partition->PartitionNumber;
+			break;
+		}
+	}
+	IFFALSE_GOTOERROR(efiPartitionNumber != -1, "ESP not found.");
+	// Fail if EFI partition number is bigger than we can fit in the
+	// uin8_t that AltMountVolume receives as parameter for partition number
+	IFFALSE_GOTOERROR(efiPartitionNumber <= 0xFF, "EFI partition number is bigger than 255.");
+
+	*espMountLetter = AltMountVolume(ConvertUnicodeToUTF8(systemDriveLetter.Left(2)), (uint8_t)efiPartitionNumber);
+	IFFALSE_GOTOERROR(*espMountLetter != NULL, "Error assigning a letter to the ESP.");
+
+	retResult = TRUE;
+error:
+	return retResult;
+}
+
+// A combination of 2 of the options found here: http://www.catch22.net/tuts/self-deleting-executables
+// This needs to be tested on all supported OS versions; I tested on Windows 8 and XP and it works
+void CEndlessUsbToolDlg::DelayDeleteFolder(const CString &folder)
+{
+	wchar_t *shortPath = NULL, cmd[MAX_PATH], params[MAX_PATH];
+	int timeoutSeconds = 5;
+
+	long  result = GetShortPathName(folder, shortPath, 0);
+	IFFALSE_GOTOERROR(result != 0, "Error on getting size needed for GetShortPathName");
+	shortPath = new wchar_t[result];
+
+	result = GetShortPathName(folder, shortPath, result);
+	IFFALSE_RETURN(result != 0, "Error on GetShortPathName");
+
+	// cmd.exe /q /c "for /l %N in () do timeout 5 >> NUL & @rmdir /S /Q C:\endless >> NUL & if not exist C:\endless exit"
+	swprintf_s(params, L"/q /c \"for /l %%N in () do timeout %d >> NUL & @rmdir /S /Q %ls >> NUL & if not exist %ls exit\"", timeoutSeconds, shortPath, shortPath);
```

Comment:
I guess the printf implementation here doesn't support the `%2$ls` syntax to select arguments by number?